crates/core/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+mod engine;
+
+use quickjs_wasm_rs::{messagepack, Context};
+
+use messagepack::{transcode_input, transcode_output};
+use once_cell::sync::OnceCell;
+use std::alloc::{alloc, dealloc, Layout};
+use std::io::{self};
+use std::ptr::copy_nonoverlapping;
+
+#[cfg(not(test))]
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+static mut JS_CONTEXT: OnceCell<Context> = OnceCell::new();
+static SCRIPT_NAME: &str = "script.js";
+
+// Unlike C's realloc, zero-length allocations need not have
+// unique addresses, so a zero-length allocation may be passed
+// in and also requested, but it's ok to return anything that's
+// non-zero to indicate success.
+const ZERO_SIZE_ALLOCATION_PTR: *mut u8 = 1 as _;
+
+/// Sets up the JS context for the engine to be used to run JS code
+///
+/// # Safety
+///
+/// TODO: When can this be called safely?
+#[export_name = "init_engine"]
+pub unsafe extern "C" fn init_engine() {
+    let mut context = Context::default();
+    context
+        .register_globals(io::stderr(), io::stderr())
+        .unwrap();
+    JS_CONTEXT.set(context).unwrap();
+}
+
+/// Evaluates the JS source code
+///
+/// # Safety
+///
+/// TODO: When can this be called safely?
+#[export_name = "init_src"]
+pub unsafe extern "C" fn init_src(js_str_ptr: *mut u8, js_str_len: usize) {
+    // TODO: Who is supposed to own this pointer? Is it the caller who allocated, or this module?
+    let js = String::from_utf8(Vec::from_raw_parts(js_str_ptr, js_str_len, js_str_len)).unwrap();
+    let context = JS_CONTEXT.get().unwrap();
+    let _ = context.eval_global(SCRIPT_NAME, &js).unwrap();
+}
+
+/// Executes the JS code.
+/// func_obj_path is expected to be a dot spearate path to the main function.
+///
+/// # Safety
+///
+/// TODO: When can this be called safely?
+#[export_name = "execute"]
+pub unsafe extern "C" fn execute(
+    func_obj_path_is_some: u32,
```

Comment:
I'm out of the loop on this, what is this parameter supposed to represent? I can see how it's used below to say use the path pointer and length if this is non-zero and use `"Shopify.main"` otherwise. Is this parameter something proprietary to Shopify or something more standard?

Reply:
If this is Shopify specific, I guess as a follow-up nit, can we change this to `func_obj_path_is_defined` or `func_obj_path_is_provided`? I get where `is_some` is coming from but it initially confused me.

Reply:
This is how `option<string>` in `.wit` is handled by the canonical ABI. It is translated to `(u32, u32, u32)` where the first argument indicates if the argument is present or not. `..._is_some` is used for consistency with rusts `Option<...>` type.

This argument suffix we use is never exposed to actual consumers of the library, they would see something in `.wit` like `execute: function(func-obj-path: option<string>)` and the actual webassembly file does not have named arguments.